staging/src/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go

```diff
@@ -100,7 +100,8 @@ func makeUsefulPanic(v reflect.Value) {
 // Tests for deep equality using reflected types. The map argument tracks
 // comparisons that have already been seen, which allows short circuiting on
 // recursive types.
-func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool, depth int) bool {
+// equateNilAndEmpty controls whether empty maps/slices are equivalent to nil
+func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool, equateNilAndEmpty bool, depth int) bool {
```

Comment:
Visited could be a `map[visit]struct{}`, that will make it smaller but probably not reduce the number of allocations

Reply:
Why does this function have both depth and a visited map? Is depth actually used?

Reply:
It appears as if `depth` is only ever `+1`'d for recursive calls

Reply:
Why does the `visit` type include the type and not just the pointers?

Reply:
looks like there have been optimizations applied since we forked this in 2017:

* https://github.com/golang/go/commit/c4f9fa1f8ee623558ed40d4d4cd3c616697cc77b
* https://github.com/golang/go/commit/a0c409cbc82ea9999a03fa0bfe6d9a8953e780e0

also, bugfixes:

* https://github.com/golang/go/commit/4dc11ae26b3a092bbcd1b4d0ae8cdaeb1cf1cef9
* https://github.com/golang/go/commit/d872bbcfec29f2585ed8144fba6262d4228619cf
* https://github.com/golang/go/commit/009d71409821a6ac4f1b32aaae2c856c20a29f92

Reply:
it looks like we also have two copies of this... third_party/forked/golang/reflect/deep_equal.go and staging/src/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go

outside the scope of this PR, but we should have at most one, and should be paying attention to what lands upstream to pick up bug/perf fixes

for this PR, can we at least avoid diverging our two copies so that we can collapse onto one in the future?

Reply:
We we know if `third_party/forked/golang/reflect` is used anywhere? A search of k8s codebase via `rg` makes it seem unused?

```
src/k8s.io/kubernetes
❯ rg third_party/forked/golang/reflect 
vendor/modules.txt
1419:k8s.io/apimachinery/third_party/forked/golang/reflect
1927:k8s.io/code-generator/third_party/forked/golang/reflect

staging/src/k8s.io/apimachinery/pkg/conversion/deep_equal.go
20:	"k8s.io/apimachinery/third_party/forked/golang/reflect"

staging/src/k8s.io/code-generator/cmd/go-to-protobuf/protobuf/parser.go
33:	customreflect "k8s.io/code-generator/third_party/forked/golang/reflect"
```

(the customreflect package at the end is a fork of only `types.go`: https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/code-generator/third_party/forked/golang/reflect)

Reply:
delete it and see if everything still compiles...

Reply:
Done in https://github.com/kubernetes/kubernetes/pull/109244, just pending master reopen

Reply:
I had already submitted a PR for its removal. It is being removed in 1.25 milestone here...